### ISSUE TITLE
Feign OkhttpClient doesn't add Content-Type as header when body is null for GET request

### ISSUE THREAD

**Issue by gauravrmazra** [Netflix/feign#391]
I am facing issue with feign.OkHttpClient. The problem is if the request is of GET type and request body is null, then Content-Type header is not added in the requestBody and also not in headers.
<strike>This creates a problem where server accepts only spectific type of Content.</strike>
This creates a problem when Server consumes and/or accepts specific type of content.

class is feign.okhttp.OkHttpClient.java line number 66 and 89

**Comment by adriancole** [comment]
It is a problem when the server api is screwed up and looks for
"content-type" when it should be looking for "accept" :)

Content-Type describes what *you* are sending, not what it is returning. If
you are sending null, you really aren't sending anything. Can you see if
your server accepts the correct header for this? ex. "Accept:
application/json"

-A

**Comment by gauravrmazra** [comment]
Yes, server is accepting the correct header.
I am using @RequestMapping which has produces and consumes whose value is
application/json

Now if header Content-Type is not set, it fails.

**Comment by gauravrmazra** [comment]
Server accept and consume particular type of content

**Comment by adriancole** [comment]
What a server accepts or consumes shouldn't be of issue to feign, since it
is a client. Sounds like you are using annotations on both sides of the
fence.

Can you validate what you are doing is actually correct in
spring-cloud-netflix? I don't want us to support an antipattern (sending
Content-Type when you mean Accept) unless there's sufficient cause to.

Maybe this is related to this:
https://github.com/spring-cloud/spring-cloud-netflix/issues/659

**Comment by gauravrmazra** [comment]
@adriancole  I have updated my question a bit to be more clear what I want to say.

**Comment by gauravrmazra** [comment]
What about DELETE request which can have RequestBody. In feign.okHttp.OkHttpClient at line 82, it handles null body for POST and PUT request but not for DELETE. So shouldn't it have same sort of handling for DELETE request.

**Comment by adriancole** [comment]
there's no defined semantics for DELETE with a body, and as such we'd not
encourage using this anti-pattern with special handling.

http://tools.ietf.org/html/rfc7231#section-4.3.5

**Comment by adriancole** [comment]
In general, it sounds like you have a very awkward api, one that
expects to send Content-Type, where http usually needs Accept, and
also DELETE with body. Until this becomes a very popular issue, I'm
going to close it as won't fix.

When doing things like this, the best path is to write your own
implementation of Client. I've had to do this myself with poorly
designed apis that use DELETE where POST should have been used.

Basically, your Client can do things like switching from Accept to
Content-Type header or otherwise, in order to satisfy an awkward
server.

**Comment by matthiasa** [comment]
Hi adriancole,
we had the same problem with jaxrs (jboss resteasy impl). Please notice the documentation:
https://docs.jboss.org/resteasy/docs/2.2.0.GA/userguide/html/JAX-RS_Content_Negotiation.html
There they write, that they use the "Content-Type" header to decide.:
> When a client makes a request, JAX-RS first finds all methods that match the path, then, it sorts things based on the content-type header sent by the client. 
Our solution was to patch the feign.okhttp.OkHttpClient, that it writes the content-type header into the request. 

`for (String value : input.headers().get(field)) {
                if (field.equalsIgnoreCase("Content-Type")) {
                    mediaType = MediaType.parse(value);
                    if (input.charset() != null) {
                        mediaType.charset(input.charset());
                    }
                }
                requestBuilder.addHeader(field, value);
            }`

Maybe you could have another look onto this issue?

Best Regards,
Matthias

**Comment by spencergibb** [comment]
This was closed as won't fix

**Comment by farukciber** [comment]
In case somebody still has problems sending GET request with Feign OkHttp with 415 error, there is a workaround. Add an interceptor to the OkHttp builder.
`new OkHttpClient.Builder().addInterceptor(new Interceptor()).build();`.

Then create an Interceptor class as such: 

```
public class Interceptor implements okhttp3.Interceptor {

    @NotNull
    @Override
    public Response intercept(okhttp3.Interceptor.Chain chain) throws IOException {
        Request originalRequest = chain.request();
        Request requestWithUserAgent = originalRequest
                .newBuilder()
                .header("Content-Type", "application/json")
                .build();

        return chain.proceed(requestWithUserAgent);
    }
}
```
